In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error 

In [2]:
#Task 1. Prepare the dataset

df = pd.read_csv('student-scores.csv')
print(df.head())

print(f"\nrows: {df.shape[0]}, columns: {df.shape[1]}")
print(f"\nmissing values: \n{df.isnull().sum()}")
print(f"\ndatatypes: \n{df.dtypes}")
df=df.dropna()

df = df.drop(columns=['first_name', 'last_name', 'email'])

categorical_cols=df.select_dtypes(exclude=['int64','float64']).columns.tolist()
for col in categorical_cols:
    df[col]=df[col].astype('category').cat.codes

print(f"\nData after cleaning: \n{df.dtypes}")
print(f"\ntotal missing values: \n{df.isnull().sum().sum()}")


   id first_name last_name                                  email  gender  \
0   1       Paul     Casey         paul.casey.1@gslingacademy.com    male   
1   2   Danielle  Sandoval  danielle.sandoval.2@gslingacademy.com  female   
2   3       Tina   Andrews       tina.andrews.3@gslingacademy.com  female   
3   4       Tara     Clark         tara.clark.4@gslingacademy.com  female   
4   5    Anthony    Campos     anthony.campos.5@gslingacademy.com    male   

   part_time_job  absence_days  extracurricular_activities  \
0          False             3                       False   
1          False             2                       False   
2          False             9                        True   
3          False             5                       False   
4          False             5                       False   

   weekly_self_study_hours   career_aspiration  math_score  history_score  \
0                       27              Lawyer          73             81   
1         

In [3]:
#Task 2. Define Target and Features (part D: Use biology_score as the target variable)

y=df['biology_score']

#input features
X=df[['math_score','physics_score','weekly_self_study_hours','absence_days']]
print(f"\nTarget variable: \n{y.describe()}")
print(f"\nSelected features: \n{X.head()}")


Target variable: 
count    2000.00000
mean       79.58150
std        13.72219
min        30.00000
25%        69.00000
50%        81.00000
75%        91.00000
max       100.00000
Name: biology_score, dtype: float64

Selected features: 
   math_score  physics_score  weekly_self_study_hours  absence_days
0          73             93                       27             3
1          90             96                       47             2
2          81             95                       13             9
3          71             88                        3             5
4          84             65                       10             5


## Justification:

1. math_score:
Math and biology both require strong analytical and logical thinking.
Students who perform well in math tend to have strong study skills
that transfer to other subjects including biology.
Therefore math score is a relevant predictor of biology performance.

2. physics_score:
Physics and biology are both natural science subjects.
Students who understand scientific concepts in physics
are more likely to grasp biological concepts as well.
Science subjects are strongly correlated with each other,
making physics score a useful feature for predicting biology score.

3. weekly_self_study_hours:
The more time a student spends studying independently each week,
the better their understanding of all subjects including biology.
Self-study hours directly reflect a student's motivation and effort,
which are key factors in academic performance.
A student who studies more hours per week is more likely
to achieve a higher biology score.

4. absence_days:
The more days a student is absent from school,
the more lessons and material they miss.
Missing classes directly reduces a student's knowledge
and negatively affects their scores in all subjects.
Absence days is therefore a strong negative predictor —
higher absence means lower expected biology score.

In [4]:
#Task 3. Split the dataset

X_train, X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)
print("\nDataset split:")
print(f"\nTotal samples: {len(X)}")
print(f"\nTraining samples: {len(X_train)}")
print(f"\nTesting samples: {len(X_test)}")


Dataset split:

Total samples: 2000

Training samples: 1600

Testing samples: 400


In [5]:
#Task 4. Train Regression Model

#model 1 and 2
model1=LinearRegression()
model1.fit(X_train,y_train)
print("\nmodel 1 trained successfully")

model2=DecisionTreeRegressor(random_state=42)
model2.fit(X_train,y_train)
print("\nmodel 2 trained successfully")


model 1 trained successfully

model 2 trained successfully


In [6]:
#Task 5. Make predictions

y_pred1=model1.predict(X_test)
y_pred2=model2.predict(X_test)

comparison=pd.DataFrame({'Actual': y_test.values, 'Linear Regression Predicted': y_pred1.round(2), 'Decision Tree Predicted': y_pred2.round(2)})
print(comparison.head(20))

    Actual  Linear Regression Predicted  Decision Tree Predicted
0       96                        81.21                     63.0
1       68                        77.30                     69.0
2       75                        77.99                     60.0
3       63                        78.14                     66.0
4       69                        77.70                     67.0
5       90                        78.51                     63.0
6      100                        79.77                     89.0
7      100                        81.40                     72.0
8       76                        76.08                     68.0
9       91                        78.94                     77.0
10      87                        80.08                     84.0
11      78                        76.68                     68.0
12      82                        76.05                     78.0
13      78                        82.53                     73.0
14      69               

In [7]:
#Task 6. Evaluate Models

#rmse = root mean squared error
#r^2 = r-squared score
#mae = mean absolute error
rmse1=np.sqrt(mean_squared_error(y_test, y_pred1))
r2_1=r2_score(y_test,y_pred1)
mae1=mean_absolute_error(y_test,y_pred1)

rmse2=np.sqrt(mean_squared_error(y_test,y_pred2))
r2_2=r2_score(y_test, y_pred2)
mae2=mean_absolute_error(y_test,y_pred2)

print("\nModel evaluation")
print("\nModel 1. Lenear Regression")
print(f"\nRMSE: {rmse1:.4f}, \nR^2: {r2_1:.4f}, \nMAE: {mae1:.4f}")
print("\nModel 2. Decision Tree Regressor")
print(f"\nRMSE: {rmse2:.4f}, \nR^2: {r2_2:.4f}, \nMAE: {mae2:.4f}")


Model evaluation

Model 1. Lenear Regression

RMSE: 14.0874, 
R^2: 0.0560, 
MAE: 11.3379

Model 2. Decision Tree Regressor

RMSE: 20.2798, 
R^2: -0.9563, 
MAE: 15.9100


In [8]:
#Task 7. Model Comparison

comparison_table=pd.DataFrame({
    'Metric': ['RMSE', 'R^2', 'MAE'], 
    'Linear Regression': [round(rmse1, 4), round(r2_1, 4), round(mae1, 4)], 
    'Decision Tree' : [round(rmse2, 4), round(r2_2, 4), round(mae2, 4)]
})

print("Model Comparison Table")
print(comparison_table)

if rmse1 < rmse2 and mae1 < mae2:
    print("Linear Regression performs better (lower RMSE and MAE)")
elif rmse2 < rmse1 and mae2 < mae1:
    print("Decision Tree performs better (lower RMSE and MAE)")
else:
    print("Models have mixed performance depending on the metric")

Model Comparison Table
  Metric  Linear Regression  Decision Tree
0   RMSE            14.0874        20.2798
1    R^2             0.0560        -0.9563
2    MAE            11.3379        15.9100
Linear Regression performs better (lower RMSE and MAE)


## Task 7. Which model is better?

1. WHICH MODEL PERFORMS BETTER? 
Linear Regression performs better across all three metrics.

RMSE (Root Mean Squared Error):
Linear Regression RMSE = 14.09, Decision Tree RMSE = 20.28.
RMSE shows the average error in the same units as the target variable — biology scores.
This means Linear Regression is wrong by about 14 points on average,
while Decision Tree is wrong by about 20 points on average.
Lower RMSE = fewer errors = better model.
Linear Regression makes significantly smaller mistakes.

R² (R-squared):
Linear Regression R² = 0.056, Decision Tree R² = -0.956.
R² shows how well the model explains the patterns in the data.
A value of 1.0 means perfect prediction.
A value of 0.0 means the model is no better than always predicting the average.
A NEGATIVE value means the model is even WORSE than just predicting the average every time.
Linear Regression explains 5.6% of the data patterns — which is weak,
but Decision Tree has a negative R², meaning it completely failed
to learn any useful patterns from the data.

MAE (Mean Absolute Error):
Linear Regression MAE = 11.34, Decision Tree MAE = 15.91.
MAE shows the average absolute difference between predicted and actual values.
Linear Regression is wrong by 11 points on average,
Decision Tree is wrong by almost 16 points on average.
Lower MAE = predictions are closer to real values = better model.

2. WHICH MODEL HAS LOWER ERROR? 
Linear Regression has lower error in all three metrics:
- RMSE is 6 points lower than Decision Tree
- MAE is 4.5 points lower than Decision Tree
- R² is significantly higher than Decision Tree

Overall conclusion on comparison:
Linear Regression is clearly the better model for this dataset.
Decision Tree completely overfitted — it memorized the training data but failed to generalize to new unseen test data, which is why its R² is negative.

## Task 8. Analysis

1. Which model performed better?
Linear Regression performed better than Decision Tree Regressor.
It achieved lower RMSE (14.09 vs 20.28), lower MAE (11.34 vs 15.91),
and higher R² (0.056 vs -0.956) across all evaluation metrics.
This means Linear Regression made smaller prediction errors
and was more consistent when predicting biology scores.

2. Why do you think this model performed better?
Decision Tree Regressor suffered from overfitting.
Overfitting means the model learned the training data too well —
it memorized specific patterns and noise in the training set,
but when it saw new unseen data (test set), it failed completely.
This is why its R² is negative — it performed worse than
simply predicting the average biology score every time.
Linear Regression is a simpler model with fewer parameters,
which makes it less likely to overfit on this dataset.
Simpler models often generalize better to new data
when the dataset does not have strong non-linear patterns.

3. How does the target variable affect model performance?
Biology score as a target variable ranges from 30 to 100
with a mean of 79.58 and standard deviation of 13.72.
This wide spread of values makes it harder for models to predict accurately.
The selected features — math score, physics score,
study hours and absence days — have a weak relationship
with biology score, which explains why both models
show relatively poor performance overall.
If features like chemistry_score or history_score were included,
the models might have performed better since these subjects
are more directly related to biology.

## Conclusion

In this practice, I built and evaluated two regression models
to predict student biology scores using four input features:
math score, physics score, weekly study hours and absence days.
Linear Regression outperformed Decision Tree Regressor across
all three metrics — RMSE, MAE and R² — proving to be
the more reliable model for this prediction task.
The most challenging part was interpreting the evaluation metrics,
especially understanding why a negative R² means the model
performed worse than simply predicting the average value every time.
Both models showed weak overall performance, which suggests
that the selected features do not have a strong enough relationship
with biology score, and future work could explore additional
or different features to improve prediction accuracy.
The dataset was successfully prepared, both models were trained
and evaluated, completing the full ML regression workflow.